File: algorithms/medium/partition_equal_subset_sum.ex
defmodule Solution do
  @spec can_partition(nums :: [integer]) :: boolean
  def can_partition(nums) do
    sum = Enum.sum(nums)

    if rem(sum, 2) != 0 do
      false
    else
      target = div(sum, 2)

      for n <- nums, s <- target..n, reduce: %{0 => true} do
        dp -> Map.put(dp, s, Map.get(dp, s, false) || Map.get(dp, s - n, false))
      end
      |> Map.get(target)
    end
  end
end

File: algorithms/medium/partition_equal_subset_sum.rs
impl Solution {
    pub fn can_partition(nums: Vec<i32>) -> bool {
        let mut sum = nums.iter().sum::<i32>() as usize;
        
        if sum % 2 != 0 {
            return false;
        }

        sum /= 2;
        
        let mut dp = vec![false; sum + 1];
        
        dp[0] = true;
        
        for n in nums {
            for j in (n as usize..=sum).rev() {
                dp[j] = dp[j] || dp[j - n as usize];
            }
        }

        dp[sum]
    }
}
